iwyu_namespace.cc

```diff
@@ -0,0 +1,249 @@
+
+  // Get the intermediate map, creating it if necessary
+  map<string, IwyuNsFileInfo*>& m = nsfileinfo[ns_ident];
+
+  if (auto pair = m.find(filepath); pair != m.end()) {
+    fileinfo = pair->second;
+  } else {
+    // filepath not known for this namespace. Create an
+    // IwyuNsFileInfo for it
+    fileinfo = new IwyuNsFileInfo(ns, filepath);
+    m[filepath] = fileinfo;
+  }
+
+  return fileinfo;
+}
+
+static const NamespaceDecl* getBestNamespaceDecl(const NamespaceDecl* ns) {
+  string ns_ident = string(ns->getQualifiedNameAsString());
+  auto m = nsfileinfo[ns_ident];
+
+  const NamespaceDecl* best_decl = nullptr;
+  const NamespaceDecl* decl = nullptr;
+  int full_decl_count = 0;
+  bool have_full_decls = false;
+  bool multiple_files = false;
+
+  for (auto& [filepath, fileinfo] : m) {
+    decl = fileinfo->GetNamespaceDecl();
+    unsigned count = fileinfo->GetFullDeclCount();
+
+    if (count > 0) {
+      if (!have_full_decls) {
+        have_full_decls = true;
+        full_decl_count = count;
+        best_decl = decl;
+      } else {
+        multiple_files = true;
+        full_decl_count += count;
+        best_decl = nullptr;
+      }
+    }
+  }
+
+  if (have_full_decls && !multiple_files) {
+    int line = GetLineNumber(best_decl->getBeginLoc());
+    VERRS(7) << "Single namespace for " << ns_ident << " "
+             << GetFilePath(best_decl) << " line " << line << " with "
+             << full_decl_count << " full decls\n";
+    return best_decl;
+  }
+
+  if (decl == nullptr) {
+    // No declarations recorded at all. Return the input NamespaceDecl
+    int line = GetLineNumber(ns->getBeginLoc());
+    VERRS(7) << "No declarations in namespace " << ns_ident << " "
+             << GetFilePath(ns) << " line " << line << "\n";
+    return ns;
+  }
+
+  if (!have_full_decls) {
+    // No full declarations. Return the input NamespaceDecl
+    int line = GetLineNumber(ns->getBeginLoc());
+    VERRS(7) << "No namespace with a full decl for " << ns_ident << " "
+             << GetFilePath(ns) << " line " << line << "\n";
+    return ns;
+  }
+
+  VERRS(7) << "Multiple namespace decls for " << ns_ident << " with "
+           << full_decl_count << " full decls. Ignoring namespace use\n";
+  return nullptr;
+}
+
+void RecordDeclNamespaces(const NamedDecl* decl) {
+  const NamespaceDecl* ns;
+
+  // Ignore declarations of sub-namespaces i.e. don't count them as a
+  // forward decl or full decl.
+  if ((ns = DynCastFrom(decl)))
+    return;
+
+  const DeclContext* decl_context = decl->getDeclContext();
+
+  bool fwd_decl = IsForwardDecl(decl);
+
+  for (; decl_context; decl_context = decl_context->getParent()) {
+    // Stop traversing up the context - function arguments, structure
+    // members and variables in block scope don't need to be recorded,
+    // as the function/structure will be.
+    //
+    // TODO: others?
+    if (const clang::FunctionDecl* func = DynCastFrom(decl_context)) {
```

Comment:
Are these boundary conditions necessary? In the majority of cases, I think there might be a `RecordDecl` at most between these and the `TranslationUnitDecl`, so if it's an optimization to exit early, I don't think it's going to buy us very much.

Or does it have any functional impact?

Reply:
This has a functional impact on the count of declarations made within a namespace. `RecordNamespaceDecls()` is being called for at least function arguments and structure fields. I believe it's getting called for variable declarations within function scope too.

So these aborts ensure that a function within a namespace only results in an increment of 1 (on all parent namespaces), similarly a struct.

Things _may_ work ok with these removed, but it's easier to validate the counts in the error log if they conform to what you'd expect.

Reply:
Side note: you can replace the three `DynCastFrom` with `isa<FunctionDecl, TagDecl, BlockDecl>(decl_context)`

Reply:
Yes, that's right.